retroshare-gui/src/gui/common/FriendListModel.h

```diff
@@ -0,0 +1,226 @@
+ *                                                                             *
+ *******************************************************************************/
+
+#pragma once
+
+#include <QModelIndex>
+#include <QColor>
+
+#include "retroshare/rsstatus.h"
+#include "retroshare/rsmsgs.h"
+#include "retroshare/rspeers.h"
+
+typedef uint32_t ForumModelIndex;
+
+// This class is the item model used by Qt to display the information
+
+class RsFriendListModel : public QAbstractItemModel
+{
+	Q_OBJECT
+
+public:
+	explicit RsFriendListModel(QObject *parent = NULL);
+	~RsFriendListModel(){}
+
+    class RsNodeDetails: public RsPeerDetails {};// in the near future, there will be a specific class for Profile/Node details in replacement of RsPeerDetails
+    class RsProfileDetails: public RsPeerDetails {};
+
+	struct HierarchicalGroupInformation
+	{
+        RsGroupInfo group_info;
+		std::vector<uint32_t> child_profile_indices;	// index in the array of hierarchical profiles
+	};
+	struct HierarchicalProfileInformation
+	{
+        RsProfileDetails profile_info;
+		std::vector<uint32_t> child_node_indices;		// indices of nodes in the array of nodes.
+	};
+	struct HierarchicalNodeInformation
+	{
+        HierarchicalNodeInformation() : last_update_ts(0) {}
+
+        rstime_t last_update_ts;
+        RsNodeDetails node_info;
+	};
+
+	enum Columns {
+		COLUMN_THREAD_NAME         = 0x00,
+		COLUMN_THREAD_LAST_CONTACT = 0x01,
+		COLUMN_THREAD_IP           = 0x02,
+		COLUMN_THREAD_ID           = 0x03,
+		COLUMN_THREAD_NB_COLUMNS   = 0x04
+	};
+
+	enum Roles{ SortRole           = Qt::UserRole+1,
+              	StatusRole         = Qt::UserRole+2,
+              	UnreadRole         = Qt::UserRole+3,
+              	FilterRole         = Qt::UserRole+4,
+              	OnlineRole         = Qt::UserRole+5
+              };
+
+    enum FilterType{ FILTER_TYPE_NONE = 0x00,
+        			 FILTER_TYPE_ID   = 0x01,
+                     FILTER_TYPE_NAME = 0x02
+                   };
+
+    enum EntryType{ ENTRY_TYPE_UNKNOWN   = 0x00,
+                    ENTRY_TYPE_GROUP     = 0x01,
+                    ENTRY_TYPE_PROFILE   = 0x02,
+                    ENTRY_TYPE_NODE      = 0x03
+                  };
+
+    // This structure encodes the position of a node in the hierarchy. The type tells which of the index fields are valid.
+
+    struct EntryIndex
+    {
+    public:
+        EntryIndex() : type(ENTRY_TYPE_UNKNOWN),group_index(0xff),profile_index(0xff),node_index(0xff) {}
+
+        EntryType type;		        // type of the entry (group,profile,location)
+
+        // Indices w.r.t. parent. The set of indices entirely determines the position of the entry in the hierarchy.
+        // An index of 0xff means "undefined"
+
+        uint8_t group_index;		// index of the group in mGroups tab
+        uint8_t profile_index;		// index of the child profile in its own group if group_index < 0xff, or in the mProfiles tab otherwise.
```

Comment:
Will this (8bit integer) work with many friends / many locations?

Reply:
very relevant question. As it is now, it limits to 256 profiles with 256 locations each into 256 groups. Do you think this is enough? If not, we can pack these onto a larger number such as uint64_t, and the problem will only concern 32bits implementations.

Reply:
It should be enough for groups and locations but could be a problem for people with many friends. 256 friends is surely a lot but nothing impossible.

Reply:
done

Reply:
nice solution!
This way we can tell people to upgrade their system/OS when they complain :)

Though you forgot to update the actual definition of `group_index` and the others (i can't see it in the diff).